python/ray/_private/runtime_env/packaging.py

```diff
@@ -393,34 +434,140 @@ def download_and_unpack_package(
+                            "`pip install boto3` to fetch URIs in s3 "
+                            "bucket.")
+                    tp = {"client": boto3.client("s3")}
+                else:
+                    try:
+                        from smart_open import open
+                    except ImportError:
+                        raise ImportError(
+                            "You must `pip install smart_open` "
+                            f"to fetch {protocol.value.upper()} URIs.")
+
                 with open(pkg_uri, "rb", transport_params=tp) as package_zip:
                     with open(pkg_file, "wb") as fin:
                         fin.write(package_zip.read())
+
+                unzip_package(
+                    package_path=pkg_file,
+                    target_dir=local_dir,
+                    remove_top_level_directory=True,
+                    unlink_zip=True,
+                    logger=logger)
             else:
                 raise NotImplementedError(
                     f"Protocol {protocol} is not supported")
 
-            os.mkdir(local_dir)
-            logger.debug(f"Unpacking {pkg_file} to {local_dir}")
-            with ZipFile(str(pkg_file), "r") as zip_ref:
-                zip_ref.extractall(local_dir)
-            pkg_file.unlink()
-
         return str(local_dir)
 
 
+def get_top_level_dir_from_compressed_package(package_path: str):
+    """
+    If compressed package at package_path contains a single top-level
+    directory, returns the name of the top-level directory. Otherwise,
+    returns None.
+    """
+
+    package_zip = ZipFile(package_path, "r")
+    top_level_directory = None
+
+    for file_name in package_zip.namelist():
+        if top_level_directory is None:
+            # Cache the top_level_directory name when checking
+            # the first file in the zipped package
+            if "/" in file_name:
+                top_level_directory = file_name.split("/")[0]
+            else:
+                return None
+        else:
+            # Confirm that all other files
+            # belong to the same top_level_directory
+            if "/" not in file_name or \
+                    file_name.split("/")[0] != top_level_directory:
+                return None
+
+    return top_level_directory
+
+
+def extract_file_and_remove_top_level_dir(base_dir: str, fname: str,
+                                          zip_ref: ZipFile):
+    """
+    Extracts fname file from zip_ref zip file, removes the top level directory
+    from fname's file path, and stores fname in the base_dir.
+    """
+
+    fname_without_top_level_dir = "/".join(fname.split("/")[1:])
+
+    # If this condition is false, it means there was no top-level directory,
+    # so we do nothing
+    if fname_without_top_level_dir:
+        zip_ref.extract(fname, base_dir)
+        os.rename(
+            os.path.join(base_dir, fname),
+            os.path.join(base_dir, fname_without_top_level_dir))
```

Comment:
Ah, are you trying to support both cases where we have a top level directory and don't have one here? I think we should probably just limit usage to _only_ supporting having a single TLD. Otherwise this is going to get messy. For example, we can't distinguish between when the user wanted their working_dir to have a single directory within it...

Reply:
My intention is to support only packages with a top level directory for remote URIs. I agree that it's not a good idea to support both packages with TLDs and without them for remote URIs.

In the highlighted ``extract_file_and_remove_top_level_dir`` function, the if-statement makes the function a no-op when the user uploads a compressed but empty TLD. Realistically, this is an edge case that could be undefined (why would a user upload an empty TLD– this is like uploading an empty ``working_dir`` or ``py_module``?).

However, since a zipped but empty TLD is still a valid package structure, I handle it by making the function a no-op. For developer readability, though, it might be worth making this edge case invalid and throwing an error anyways. I prefer the no-op because it's a bit more predictable and consistent than having an extra edge case, but I want to make sure readability isn't too negatively impacted. How do you think we should proceed?

Reply:
We should also wrap the ``shutil.rmtree()`` call in ``unzip_package()`` to make sure that it isn't called when ``extract_file_and_remove_top_level_dir()`` is a no-op. I'll make that change and push the code.

Reply:
We should definitely support the case when the directory inside the package is empty (users might do this accidentally and don't want to throw a very obscure error when it happens).

So just to be 100% sure, you're saying we _only_ support the case when the structure is a zip file containing a single TLD, and error when it's a zip file with multiple top-level files or directories? And the code here is handling the case where the single TLD is empty. If I'm understanding correctly, then this LGTM.

Reply:
Yes, your understanding is correct. We only support the case when the structure is a zip file containing a single TLD. If the zip file does not contain a single TLD, we raise a ``ValueError`` in ``unzip_package()``. And yep, this code here is handling the case where the single TLD is empty.